sklearn/datasets/tests/test_openml.py

```diff
@@ -459,17 +459,39 @@ def test_open_openml_url_cache(monkeypatch, gzip_response):
     _monkey_patch_webbased_functions(
         monkeypatch, data_id, gzip_response)
     openml_path = sklearn.datasets.openml._DATA_FILE.format(data_id)
-    test_directory = os.path.join(os.path.expanduser('~'), 'scikit_learn_data')
+    cache_directory = os.path.join(os.path.expanduser('~'), 'scikit_learn_data')
```

Comment:
What do you think about using a tempdir for the cache directory during testing:

```python
def test_open_openml_url_cache(monkeypatch, gzip_response, tmpdir):
    ...
    cache_directory = str(tmpdir.mkdir('scikit_learn_data'))
```

This can also be used in `test_fetch_openml_cache`.

Reply:
The longer I think about it, the more important I think it is. Due to using the same data ids as the main server, yet not using the same files, developers that run unit tests can obtain files in their cache that are inconsistent with the real datasets. 

I will try to add a sensible fix

Reply:
Actually, I wasn't able to fix this. What is the tmpdir library? Everything I find in the Pydocs points to the tempfile library. The following code leads to the following error:
code
```
cache_directory = os.path.join(tempfile.mkdtemp(),
                                   'scikit_learn_test_data')
```


error
```
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fd5898692e8>, gzip_response = False

    @pytest.mark.parametrize('gzip_response', [True, False])
    def test_open_openml_url_cache(monkeypatch, gzip_response):
        data_id = 61
    
        _monkey_patch_webbased_functions(
            monkeypatch, data_id, gzip_response)
        openml_path = sklearn.datasets.openml._DATA_FILE.format(data_id)
        cache_directory = os.path.join(tempfile.mkdtemp(),
                                       'scikit_learn_test_data')
        # first fill the cache
        response1 = _open_openml_url(openml_path, cache_directory)
        # assert file exists
        location = _get_local_path(openml_path, cache_directory)
        assert os.path.isfile(location)
        # redownload, to utilize cache
        response2 = _open_openml_url(openml_path, cache_directory)
>       assert response1.read() == response2.read()

test_openml.py:478: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../../../anaconda3/envs/scikit-learn34/lib/python3.4/gzip.py:360: in read
    while self._read(readsize):
../../../../../anaconda3/envs/scikit-learn34/lib/python3.4/gzip.py:433: in _read
    if not self._read_gzip_header():
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <gzip _io.BufferedReader name='/tmp/tmp80braqj1/scikit_learn_test_data/openml.org/data/v1/download/61.gz' 0x7fd589869d68>

    def _read_gzip_header(self):
        magic = self.fileobj.read(2)
        if magic == b'':
            return False
    
        if magic != b'\037\213':
>           raise OSError('Not a gzipped file')
E           OSError: Not a gzipped file
```

Reply:
`tmpdir` is a pytest fixture: https://docs.pytest.org/en/latest/tmpdir.html. It is included when pytest is installed. For example, you can use it as follows:

```python
@pytest.mark.parametrize('gzip_response', [True, False])
 def test_fetch_openml_cache(monkeypatch, gzip_response, tmpdir):
     def _mock_urlopen_raise(request):
         raise ValueError('This mechanism intends to test correct cache'
                          'handling. As such, urlopen should never be '
                          'accessed. URL: %s' % request.get_full_url())
     data_id = 2
     cache_directory = str(tmpdir.mkdir('scikit_learn_data'))
     _monkey_patch_webbased_functions(
         monkeypatch, data_id, gzip_response)
     X_f, y_f = fetch_openml(data_id=data_id, cache=True,
                             data_home=cache_directory, return_X_y=True)

     monkeypatch.setattr(sklearn.datasets.openml, 'urlopen',
                         _mock_urlopen_raise)

     X_c, y_c = fetch_openml(data_id=data_id, cache=True,
                             data_home=cache_directory, return_X_y=True)
     np.testing.assert_array_equal(X_f, X_c)
     np.testing.assert_array_equal(y_f, y_c)
```

`tmpdir` just needs to be included as a parameter to `test_fetch_openml_cache`.

Reply:
This error is happening when `gzip=False`. When the fsrc object is returned as a non gzipped binary, it is saved as a gzip file anyways. This can be fixed by adjusting `_open_openml_url`:

```python
    if not os.path.exists(local_path):
        fsrc = urlopen(req)
        try:
            os.makedirs(os.path.dirname(local_path))
        except OSError:
            # potentially, the directory has been created already
            pass
        try:
            if is_gzip(fsrc):
                with open(local_path, 'wb') as fdst:
                    shutil.copyfileobj(fsrc, fdst)
                    fsrc.close()
            else:
                with gzip.GzipFile(local_path, 'wb') as fdst:
                    shutil.copyfileobj(fsrc, fdst)
                    fsrc.close()
```

If `fsrc` is a gzipped file, it is cached directly. If `fsrc` is not gzipped, it will be gzipped locally.

The strange thing is how the test passed when with the `os.path.expanduser('~')` directory.

Reply:
My current guess is a little bug in the `_monkey_patch_webbased_functions`. It seems to work with two notions of gzip: 1) gzip_response and 2) has_gzip_header. 

What is exactly the meaning of either?

Reply:
`has_gzip_header` is set to true when the request header includes `Accept-encoding=gzip`. `gzip_response` is provided by the tests to determine if a gzip response should be returned by the mocked urlopen. This test the case where the client request for `Accept-encoding=gzip`, but the server still decides to return a uncompressed version.

I tested this version of locally `_open_openml_url` and your new test passes:

```python
def _open_openml_url(openml_path, data_home):
    def is_gzip(_fsrc):
        return _fsrc.info().get('Content-Encoding', '') == 'gzip' 

    req = Request(_OPENML_PREFIX + openml_path)
    req.add_header('Accept-encoding', 'gzip')

    if data_home is None:
        fsrc = urlopen(req)
        if is_gzip(fsrc):
            if PY2:
                fsrc = BytesIO(fsrc.read())
            return gzip.GzipFile(fileobj=fsrc, mode='rb')
        return fsrc

    local_path = _get_local_path(openml_path, data_home)
    if not os.path.exists(local_path):
        fsrc = urlopen(req)
        try:
            os.makedirs(os.path.dirname(local_path))
        except OSError:
            # potentially, the directory has been created already
            pass

        try:
            if is_gzip(fsrc):
                with open(local_path, 'wb') as fdst:
                    shutil.copyfileobj(fsrc, fdst)
                    fsrc.close()
            else:
                with gzip.GzipFile(local_path, 'wb') as fdst:
                    shutil.copyfileobj(fsrc, fdst)
                    fsrc.close()
        except Exception:
            os.unlink(local_path)
            raise
    # XXX: First time, decompression will not be necessary (by using fsrc), but
    # it will happen nonetheless
    return gzip.GzipFile(local_path, 'rb')
```

Reply:
I just incorporated your fix and it seems to solve the problem. Thanks a lot!